src/app/ReadClient.cpp

```diff
@@ -484,14 +484,24 @@ CHIP_ERROR ReadClient::GenerateDataVersionFilterList(DataVersionFilterIBs::Build
 
 void ReadClient::OnActiveModeNotification()
 {
-    // This function just tries to complete the deferred resubscription logic in `OnLivenessTimeoutCallback`.
     VerifyOrDie(mpImEngine->InActiveReadClientList(this));
-    // If we are not in InactiveICDSubscription state, that means the liveness timeout has not been reached. Simply do nothing.
-    VerifyOrReturn(IsInactiveICDSubscription());
-
     // When we reach here, the subscription definitely exceeded the liveness timeout. Just continue the unfinished resubscription
     // logic in `OnLivenessTimeoutCallback`.
-    TriggerResubscriptionForLivenessTimeout(CHIP_ERROR_TIMEOUT);
+    if (IsInactiveICDSubscription())
+    {
+        TriggerResubscriptionForLivenessTimeout(CHIP_ERROR_TIMEOUT);
+        return;
+    }
+
+    if (mIsResubscriptionScheduled)
+    {
+        // If a subscription fails and resubscription is scheduled, a check-in message should cancel
+        // the pending resubscription and initiate a new subscription immediately."
+        mNumRetries = 0;
+        CancelResubscribeTimer();
+        TriggerResubscriptionForLivenessTimeout(CHIP_ERROR_TIMEOUT);
```

Comment:
Again: why is this not calling TriggerResubscribeIfNeeded?  This "we got a signal to resubscribe" thing is _exactly_ what that function is for.

Reply:
o, yes, that is exactly what we need! thanks